runtime/src/main/java/org/corfudb/runtime/LogReplicationListener.java

```diff
@@ -0,0 +1,182 @@
+        // If this update came before the client's full sync timestamp, ignore it.
+        if (results.getTimestamp().getSequence() <= clientFullSyncTimestamp.get()) {
+            return;
+        }
+
+        Set<String> tableNames =
+                results.getEntries().keySet().stream().map(schema -> schema.getTableName()).collect(Collectors.toSet());
+
+        if (tableNames.contains(REPLICATION_STATUS_TABLE)) {
+            Preconditions.checkState(results.getEntries().keySet().size() == 1,
+                "Replication Status Table Update received with other tables");
+            processReplicationStatusUpdate(results);
+            return;
+        }
+
+        // Data Updates
+        if (clientFullSyncPending.get()) {
+            // If the listener started when snapshot sync was ongoing, ignore all data updates until it ends.  When
+            // it ends, the client will perform a full sync and build a consistent state containing these updates.
+            return;
+        }
+
+        if (snapshotSyncInProgress.get()) {
+            processUpdatesInSnapshotSync(results);
+        } else {
+            processUpdatesInLogEntrySync(results);
+        }
+    }
+
+    private void processReplicationStatusUpdate(CorfuStreamEntries results) {
+        Map<TableSchema, List<CorfuStreamEntry>> entries = results.getEntries();
+
+        List<CorfuStreamEntry> replicationStatusTableEntries =
+            entries.entrySet().stream().filter(e -> e.getKey().getTableName().equals(REPLICATION_STATUS_TABLE))
+            .map(Map.Entry::getValue)
+            .findFirst()
+            .get();
+
+        for (CorfuStreamEntry entry : replicationStatusTableEntries) {
+
+            // Ignore any update where the operation type != UPDATE
+            if (entry.getOperation() == CorfuStreamEntry.OperationType.UPDATE) {
+                ReplicationStatusVal status = (ReplicationStatusVal)entry.getPayload();
+
+                if (status.getDataConsistent()) {
+                    // getDataConsistent() == true means that snapshot sync has ended.
+                    if (snapshotSyncInProgress.get()) {
+                        if (clientFullSyncPending.get()) {
+                            // Snapshot sync which was ongoing when the listener was subscribed has ended.  Attempt to
+                            // perform a full sync now.
+                            LogReplicationUtils.attemptClientFullSync(corfuStore, this, namespace);
+                            return;
+                        }
+                        // Process snapshot sync completion in steady state, i.e., client full sync is already complete
+                        snapshotSyncInProgress.set(false);
+                        onSnapshotSyncComplete();
+                    }
+                } else {
+                    // getDataConsistent() == false.  Snapshot sync has started.
+                    snapshotSyncInProgress.set(true);
+                    onSnapshotSyncStart();
```

Comment:
Can we reach here when snapshotSyncInProgress was already set to true? if yes, then do we need to invoke onSnapshotSyncStart again?